Pam/ch10257/update typeorm port to use latest performance

TomVance opened the pull request:
Replays the latest changes from master to this branch, and then re-applies our changes from loyaltylion/changes

clarkdave committed:
tweak package.json, etc for our fork

clarkdave committed:
add support to disable connection pooling for postgres driver

clarkdave committed:
deprecate setLock in favour of lockMode

https://github.com/typeorm/typeorm/issues/2037

clarkdave committed:
add `lock` method to SelectQueryBuilder

https://github.com/typeorm/typeorm/issues/2037

clarkdave committed:
publish 0.2.9-1

clarkdave committed:
publish 0.2.10-1

clarkdave committed:
add TransactionalEntityManager type

This type is used instead of a regular EntityManager in the
EntityManager.prototype.transaction callback. It extends an
EntityManager, so can be used as normal, but is not assignable to one,
so functions and services can require this type in order to assert they
receive a manager that has already started a transaction

clarkdave committed:
publish 0.2.10-2

clarkdave committed:
publish 0.2.10-3

atreidesend committed:
set and unset activeTransaction flag on databaseConnection

LL specific change - our legacy services use this field to determine
if a transaction is active or not, if we don't manage it via
typeorm then it's possible legacy clients will still think they
are in a transaction when they're not

atreidesend committed:
update package to 0.2.10-4

clarkdave committed:
use esnext compiler target and add ts version to vscode settings

clarkdave committed:
remove bad connections from the postgres pool after a failed query

typeorm currently lets pg.pool assume that clients are always safe to
add back into the pool, even if it encountered an error while running
a query. this is /usually/ fine, because the vast majority of errors
that happen during a query do not cause the client's connection itself
to become inoperable

however, if, during a query, the client's connection is killed, the
query will fail and the client will be placed back into the pool even
though it's linked to a now broken connection, i.e. that client will
never work again, but because pg.Pool doesn't know that, it'll keep
handing it out. end result: lots of failed queries

this patches `Connection.prototype.query` so that it passes the error
back to the driver when it calls `queryRunner.release`. the driver can
(if applicable) then inspect the error and decide if it should result
in the client being removed from the pool or just released as normal

clarkdave committed:
publish 0.2.20-1

clarkdave committed:
remove clients from the pool following any error

we're not confident in our ability to accurately determine what a "bad"
error (i.e. one that has rendered the connection broken). to be safe,
we'll instruct the pool to remove any client that failed to execute
a query

though this will result in more pool churn, we're using pgbouncer so
those connections are cheap, and this provides maximum safety overall

Commit:
Bumped node version to 13.7.0

Commit:
Bump to version 0.2.20-3

draaglom commented:
since we're forked off the latest master, our version should reflect that (0.2.29-3)

Can you add some context on the DriverUtils commit? it's not explained in the PR or the commit message

TomVance commented:
> since we're forked off the latest master, our version should reflect that (0.2.29-3)
> 
> Can you add some context on the DriverUtils commit? it's not explained in the PR or the commit message

I've removed that change, checking against master it wasn't, for some reason during the rebase it jumped from line 5 to end. Maybe some lint rule locally. I've moved it back to remove the change from the PR.

Have changed version
